Add new option --repository-file (default: $RESTIC_REPOSITORY_FILE)

mtdcr opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

<!--
Describe the changes and their purpose here, as detailed as needed.
-->
As an alternative to -r, this allows to read the repository URL from a file in order to prevent certain types of information leaks, especially for URLs containing credentials.


Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
Closes #1458
Closes #2900


Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

mtdcr committed:
Add new option --repository-file (default: $RESTIC_REPOSITORY_FILE)

As an alternative to -r, this allows to read the repository URL
from a file in order to prevent certain types of information leaks,
especially for URLs containing credentials.

Fixes #1458, fixes #2900.

rawtaz commented on cmd/restic/global.go:
```diff
@@ -101,6 +102,7 @@ func init() {
 
 	f := cmdRoot.PersistentFlags()
 	f.StringVarP(&globalOptions.Repo, "repo", "r", os.Getenv("RESTIC_REPOSITORY"), "`repository` to backup to or restore from (default: $RESTIC_REPOSITORY)")
+	f.StringVarP(&globalOptions.RepositoryFile, "repository-file", "", os.Getenv("RESTIC_REPOSITORY_FILE"), "read the repository from a `file` (default: $RESTIC_REPOSITORY_FILE)")
```
Please change `read the repository from` to `read the repository location from`, as without that it can sound like the repository is stored in the file. The wording "repository location" is used in the documentation at https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html .

rawtaz commented:
We also need a changelog file in the `changelog/unreleased` directory - see the other files in there for examples, and in particular [the template file](https://github.com/restic/restic/blob/master/changelog/TEMPLATE) for how to format the text. You can use "Enhancement" instead of "Bugfix".

mtdcr commented:
I just added two commits based on your feedback. If you prefer squashed commits, drop me a note. Or you can squash on merge.

mtdcr commented:
Btw., I'm unsure how to interpret this sentence from the template correctly: _The first issue ID (or the first PR ID, in case there aren't any issue links) is used as the primary ID._

Does this mean I should have chosen the filename based on the issue listed first? Because that's not obvious from looking at other files in changelog/unreleased (e.g. pull-2195). I chose to use the pull-request as a filename, because it fixes two (duplicated) issues. Or is this primary ID used for something else and I don't need to care about it at all?

rawtaz commented on cmd/restic/global.go:
```diff
@@ -382,15 +384,42 @@ func ReadPasswordTwice(gopts GlobalOptions, prompt1, prompt2 string) (string, er
+	repo := opts.Repo
+	if opts.RepositoryFile != "" {
+		if repo != "" {
+			return "", errors.Fatal("-r and --repository-file are mutually exclusive options")
```
Suggestion: Starting the sentence with "Option -r and --repository-file ..." might be nicer, even if it's just cosmetics.

mtdcr replied:
I like your suggestion. I should add that it was copied from `Password file and command are mutually exclusive options` in the same file, however.

rawtaz commented:
> Btw., I'm unsure how to interpret this sentence from the template correctly: `The first issue ID (or the first PR ID, in case there aren't any issue links) is used as the primary ID.`

Good question. Normally there's one issue that is the main one for tracking the problem, but in this case I've failed to notice that #2900 is a duplicate of the older #1458. I should have closed #2900 with a reference to #1458 but didn't. Then we would have referenced just #1458 in the changelog entry.

In this case, both those issues are pretty much equivalent in what they address, so let's list both of them in the changelog entry just like you have already done, and it doesn't hurt keeping the name of the changelog file referencing the PR number, since it's unique. All well! :)

MichaelEischer commented on cmd/restic/global.go:
```diff
@@ -382,15 +384,42 @@ func ReadPasswordTwice(gopts GlobalOptions, prompt1, prompt2 string) (string, er
+			return "", errors.Fatalf("%s does not exist", opts.RepositoryFile)
+		}
+
+		if err != nil {
```
please remove the extra space above.

MichaelEischer commented on cmd/restic/global.go:
```diff
@@ -101,6 +102,7 @@ func init() {
 
 	f := cmdRoot.PersistentFlags()
 	f.StringVarP(&globalOptions.Repo, "repo", "r", os.Getenv("RESTIC_REPOSITORY"), "`repository` to backup to or restore from (default: $RESTIC_REPOSITORY)")
+	f.StringVarP(&globalOptions.RepositoryFile, "repository-file", "", os.Getenv("RESTIC_REPOSITORY_FILE"), "read the repository location from a `file` (default: $RESTIC_REPOSITORY_FILE)")
```
Please add the environment variable to the corresponding list in the documentation.

MichaelEischer commented on cmd/restic/global.go:
```diff
@@ -382,15 +384,42 @@ func ReadPasswordTwice(gopts GlobalOptions, prompt1, prompt2 string) (string, er
+	}
+
+	repo := opts.Repo
+	if opts.RepositoryFile != "" {
```
Please add an integration test which tests reading the repository path from a file. (e.g. init a repository using a path stored in `Repo` then write the path to a RepositoryFile and try to access the repository again.)

MichaelEischer reviewed (commented):
Please squash the commits. I generally like commits which just do one thing, but lots of small tweaks of previous commits from the same PR aren't that useful.

mtdcr commented:
Rebased (fixed merge conflicts), squashed, addressed all comments about existing code and documentation.

But please write an integration test yourself. Go is far from being my primary language. I don't even know your test suite and I'm not willing to learn it just for a one-time patch submission.

MichaelEischer committed:
Add test for --repository-file

rawtaz commented on doc/040_backup.rst:
```diff
@@ -407,6 +407,7 @@ environment variables. The following lists these environment variables:
 
 .. code-block:: console
 
+    RESTIC_REPOSITORY_FILE              Name of a file containing the location of the repository (replaces --repository-file)
```
I'd suggest formulating this the same way as below for `RESTIC_PASSWORD_FILE`. If the reader things it's unclear what it does, they're expected to read the help for `--repository-file`.

rawtaz replied:
If you don't like the former I'd probably shorten it to "Name of file containing the repository location (replaces --repository-file)".

rawtaz commented on doc/030_preparing_a_new_repo.rst:
```diff
@@ -22,8 +22,10 @@ other options. You can skip to the next chapter once you've read the relevant
 For automated backups, restic accepts the repository location in the
-environment variable ``RESTIC_REPOSITORY``. For the password, several options
-exist:
+environment variable ``RESTIC_REPOSITORY``. restic can also read the repository
```
In the rest of the documentation "Restic" is used when starting a sentence (instead of "restic").

MichaelEischer committed:
Document new option --repository-file

MichaelEischer committed:
Update backup help output in documentation

rawtaz reviewed (approved):
Awesome work, as usual. Appr00d!
